In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('C:/SupplyChainAssistant/data/supply_chain_data.csv')
print("Rows and Columns:", df.shape)
print("\nColumn Names:")
print(df.columns.tolist())

Rows and Columns: (100, 24)

Column Names:
['Product type', 'SKU', 'Price', 'Availability', 'Number of products sold', 'Revenue generated', 'Customer demographics', 'Stock levels', 'Lead times', 'Order quantities', 'Shipping times', 'Shipping carriers', 'Shipping costs', 'Supplier name', 'Location', 'Lead time', 'Production volumes', 'Manufacturing lead time', 'Manufacturing costs', 'Inspection results', 'Defect rates', 'Transportation modes', 'Routes', 'Costs']


In [2]:
df.head()

,Product type,SKU,Price,Availability,Number of products sold,Revenue generated,Customer demographics,Stock levels,Lead times,Order quantities,...,Location,Lead time,Production volumes,Manufacturing lead time,Manufacturing costs,Inspection results,Defect rates,Transportation modes,Routes,Costs
0,haircare,SKU0,69.808006,55,802,8661.996792,Non-binary,58,7,96,...,Mumbai,29,215,29,46.279879,Pending,0.226410,Road,Route B,187.752075
1,skincare,SKU1,14.843523,95,736,7460.900065,Female,53,30,37,...,Mumbai,23,517,30,33.616769,Pending,4.854068,Road,Route B,503.065579
2,haircare,SKU2,11.319683,34,8,9577.749626,Unknown,1,10,88,...,Mumbai,12,971,27,30.688019,Pending,4.580593,Air,Route C,141.920282
3,skincare,SKU3,61.163343,68,83,7766.836426,Non-binary,23,13,59,...,Kolkata,24,937,18,35.624741,Fail,4.746649,Rail,Route A,254.776159
4,skincare,SKU4,4.805496,26,871,2686.505152,Non-binary,5,3,56,...,Delhi,5,414,3,92.065161,Fail,3.145580,Air,Route A,923.440632


In [3]:
df.dtypes

Product type                object
SKU                         object
Price                      float64
Availability                 int64
Number of products sold      int64
Revenue generated          float64
Customer demographics       object
Stock levels                 int64
Lead times                   int64
Order quantities             int64
Shipping times               int64
Shipping carriers           object
Shipping costs             float64
Supplier name               object
Location                    object
Lead time                    int64
Production volumes           int64
Manufacturing lead time      int64
Manufacturing costs        float64
Inspection results          object
Defect rates               float64
Transportation modes        object
Routes                      object
Costs                      float64
dtype: object

In [4]:
import pandas as pd
import numpy as np

# Load original - this will never be touched
df = pd.read_csv('C:/SupplyChainAssistant/data/supply_chain_data.csv')

# Working copy - all exploration and cleaning happens on this
df_clean = df.copy()

print("Original shape:", df.shape)
print("Copy shape:", df_clean.shape)
print("Both should be identical ✓")

Original shape: (100, 24)
Copy shape: (100, 24)
Both should be identical ✓


In [5]:
missing = df_clean.isnull().sum()
missing_pct = (df_clean.isnull().sum() / len(df)) * 100
missing_report = pd.DataFrame({
    'Missing Count': missing,
    'Missing Percentage': missing_pct
})
print(missing_report)

                         Missing Count  Missing Percentage
Product type                         0                 0.0
SKU                                  0                 0.0
Price                                0                 0.0
Availability                         0                 0.0
Number of products sold              0                 0.0
Revenue generated                    0                 0.0
Customer demographics                0                 0.0
Stock levels                         0                 0.0
Lead times                           0                 0.0
Order quantities                     0                 0.0
Shipping times                       0                 0.0
Shipping carriers                    0                 0.0
Shipping costs                       0                 0.0
Supplier name                        0                 0.0
Location                             0                 0.0
Lead time                            0                 0

In [7]:
print("Duplicate rows:", df_clean.duplicated().sum())

Duplicate rows: 0


In [8]:
df_clean.describe()

,Price,Availability,Number of products sold,Revenue generated,Stock levels,Lead times,Order quantities,Shipping times,Shipping costs,Lead time,Production volumes,Manufacturing lead time,Manufacturing costs,Defect rates,Costs
count,100.000000,100.000000,100.000000,100.000000,100.000000,100.000000,100.000000,100.000000,100.000000,100.000000,100.000000,100.00000,100.000000,100.000000,100.000000
mean,49.462461,48.400000,460.990000,5776.048187,47.770000,15.960000,49.220000,5.750000,5.548149,17.080000,567.840000,14.77000,47.266693,2.277158,529.245782
std,31.168193,30.743317,303.780074,2732.841744,31.369372,8.785801,26.784429,2.724283,2.651376,8.846251,263.046861,8.91243,28.982841,1.461366,258.301696
min,1.699976,1.000000,8.000000,1061.618523,0.000000,1.000000,1.000000,1.000000,1.013487,1.000000,104.000000,1.00000,1.085069,0.018608,103.916248
25%,19.597823,22.750000,184.250000,2812.847151,16.750000,8.000000,26.000000,3.750000,3.540248,10.000000,352.000000,7.00000,22.983299,1.009650,318.778455
50%,51.239831,43.500000,392.500000,6006.352023,47.500000,17.000000,52.000000,6.000000,5.320534,18.000000,568.500000,14.00000,45.905622,2.141863,520.430444
75%,77.198228,75.000000,704.250000,8253.976921,73.000000,24.000000,71.250000,8.000000,7.601695,25.000000,797.000000,23.00000,68.621026,3.563995,763.078231
max,99.171329,100.000000,996.000000,9866.465458,100.000000,30.000000,96.000000,10.000000,9.929816,30.000000,985.000000,30.00000,99.466109,4.939255,997.413450


In [9]:
print("Lead times sample:", df_clean['Lead times'].head(10).tolist())
print("Lead time sample:", df_clean['Lead time'].head(10).tolist())
print("\nAre they identical?", df_clean['Lead times'].equals(df['Lead time']))

Lead times sample: [7, 30, 10, 13, 3, 27, 15, 17, 10, 27]
Lead time sample: [29, 23, 12, 24, 5, 10, 14, 22, 13, 29]

Are they identical? False


In [10]:
categorical_cols = df_clean.select_dtypes(include='object').columns
print("Categorical columns:", categorical_cols.tolist())
print("\n")

for col in categorical_cols:
    print(f"Column: {col}")
    print(f"Unique values: {df_clean[col].nunique()}")
    print(df_clean[col].value_counts())
    print("-" * 40)

Categorical columns: ['Product type', 'SKU', 'Customer demographics', 'Shipping carriers', 'Supplier name', 'Location', 'Inspection results', 'Transportation modes', 'Routes']


Column: Product type
Unique values: 3
Product type
skincare     40
haircare     34
cosmetics    26
Name: count, dtype: int64
----------------------------------------
Column: SKU
Unique values: 100
SKU
SKU0     1
SKU63    1
SKU73    1
SKU72    1
SKU71    1
        ..
SKU30    1
SKU29    1
SKU28    1
SKU27    1
SKU99    1
Name: count, Length: 100, dtype: int64
----------------------------------------
Column: Customer demographics
Unique values: 4
Customer demographics
Unknown       31
Female        25
Non-binary    23
Male          21
Name: count, dtype: int64
----------------------------------------
Column: Shipping carriers
Unique values: 3
Shipping carriers
Carrier B    43
Carrier C    29
Carrier A    28
Name: count, dtype: int64
----------------------------------------
Column: Supplier name
Unique values: 5
S

In [11]:
for col in df_clean.columns:
    print(f"{col}: {df_clean[col].nunique()} unique values")

Product type: 3 unique values
SKU: 100 unique values
Price: 100 unique values
Availability: 63 unique values
Number of products sold: 96 unique values
Revenue generated: 100 unique values
Customer demographics: 4 unique values
Stock levels: 65 unique values
Lead times: 29 unique values
Order quantities: 61 unique values
Shipping times: 10 unique values
Shipping carriers: 3 unique values
Shipping costs: 100 unique values
Supplier name: 5 unique values
Location: 5 unique values
Lead time: 29 unique values
Production volumes: 96 unique values
Manufacturing lead time: 30 unique values
Manufacturing costs: 100 unique values
Inspection results: 3 unique values
Defect rates: 100 unique values
Transportation modes: 4 unique values
Routes: 3 unique values
Costs: 100 unique values


## Week 1 Data Exploration Summary

### Dataset Overview
- 100 rows, 24 columns
- Zero missing values, zero duplicates
- All datatypes are correct

### Cleaning Actions Needed
- Rename 'Lead times' to 'supplier_lead_time'
- Rename 'Lead time' to 'customer_lead_time'
- Standardize all column names to lowercase with underscores
- Treat 'Unknown' in Customer demographics as missing value
- Flag SKU as ID column - not useful for analysis

### Interesting Findings
- 36% product inspection failure rate - critical finding
- 41% inspections still pending
- Stock levels minimum is 0 - potential stockout exists
- Supplier 1 dominates orders with 27 out of 100
- Location data is India based - 5 cities

### Columns to Flag as ID
- SKU - 100 unique values in 100 rows

### Ready for Week 2
- Data is clean synthetically but rules are identified
- Cleaning pipeline logic is```

---

**Day 5 — Complete ✅**

**Week 1 — Complete ✅**

```Week 1 fully complete. Ready for Week 2.